crypto/pkcs7/pkcs7_x509.c

```diff
@@ -370,6 +377,100 @@ static int write_signer_info(CBB *out, const void *arg) {
   return ret;
 }
 
+static PKCS7_SIGNER_INFO *pkcs7_add_signature(PKCS7 *p7, X509 *x509,
+                                              EVP_PKEY *pkey) {
+  PKCS7_SIGNER_INFO *si = NULL;
+  const EVP_MD *digest = NULL;
+
+  EVP_PKEY_CTX *pkey_ctx = EVP_PKEY_CTX_new(pkey, /*engine*/ NULL);
+  if (!pkey_ctx) {
+    goto err;
+  }
+  int ok = EVP_PKEY_CTX_get_signature_md(pkey_ctx, &digest);
+  EVP_PKEY_CTX_free(pkey_ctx);
+  if (!ok && (EVP_PKEY_id(pkey) == EVP_PKEY_RSA ||
+              EVP_PKEY_id(pkey) == EVP_PKEY_DSA)) {
```

Comment:
...processing...

Reply:
yeah, this is fair. it's pretty subtle and caused me a lot of headache figuring out the initialization stuff. 

> I believe that a EVP_PKEY_CTX needs to be initialized for a sign or verify operation ... before EVP_PKEY_CTX_get_signature_md would return success (?)

i don't think so. here's the ctrl functions for [rsa](https://github.com/aws/aws-lc/blob/8226a0565dd81aab1c3eaf4e623301b7ce2c4423/crypto/fipsmodule/evp/p_rsa.c#L599) and [dsa](https://github.com/aws/aws-lc/blob/8226a0565dd81aab1c3eaf4e623301b7ce2c4423/crypto/evp_extra/p_dsa.c#L270). they succeed unconditionally.

actually, i think we can simplify this a bit by ripping out the call to `EVP_PKEY_CTX_get_signature_md`.

> (What about ECDSA?)

Hm, yes. It looks like OpenSSL [defines these default digests](https://github.com/search?q=repo%3Aopenssl%2Fopenssl%20ASN1_PKEY_CTRL_DEFAULT_MD_NID&type=code) for DSA, RSA and EC, so that's what we should do here.

> Only RSA/PSS initializes the digest, but RSA and DSA don't? Except for RSA/PSS, the digest remains empty until a call to EVP_PKEY_CTX_set_signature_md is made. (?)

I suppose so. RSA PSS uses [a presumably initialized `RSA_PKEY_CTX`](https://github.com/aws/aws-lc/blob/f1ff8f6fc204e682af155607a289d033f74a05e9/crypto/fipsmodule/evp/p_rsa.c#L133) and steals `md` from that, while RSA (regular) and DSA both allocate new fresh contexts.


> If initialized, the call to EVP_PKEY_CTX_get_signature_md doesn't fail even if the digest is NULL -- so the if (!ok && (...)){ ...} condition below would not provide the initialization in that case.

removed the `EVP_PKEY_CTX_get_signature_md` entirely.

> Do we have test coverage for this function?

[yes](https://app.codecov.io/gh/aws/aws-lc/pull/1993?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=aws#54f876b7f5bb1a4002cf77c45894d19b-R389).